extensions/classifiers/SVMPredictionService.js

```diff
@@ -26,109 +26,222 @@
+            var sum = 0;
+            for (var j = 0; j < input.length; j++) {
+              sum += Math.pow((supportVectors[i][j] - input[j]), 2);
+            }
+            kvalues.push(Math.exp(-gamma * sum));
+          }
+        } else if (kernel === 'linear') {
+          var vectorLength = input.length;
+          for (var i = 0; i < supportVectors.length; i++) {
+            var sum = 0;
+            for (var j = 0; j < input.length; j++) {
+              sum += supportVectors[i][j] * input[j];
+            }
+            kvalues.push(sum);
           }
-          kvalues.push(Math.exp(-gamma * sum));
         }
-      } else if (kernel === 'linear') {
-        var vectorLength = input.length;
-        for (var i = 0; i < supportVectors.length; i++) {
-          var sum = 0;
-          for (var j = 0; j < input.length; j++) {
-            sum += supportVectors[i][j] * input[j];
+        return kvalues;
+      },
+
+      multiclassProbability: function(nClasses, pairwiseProb) {
+        var Q = [];
+        for (var i = 0; i < nClasses; i++) {
+          Q.push([]);
+          for (var j = 0; j < nClasses; j++) {
+            Q[i].push(0);
           }
-          kvalues.push(sum);
         }
-      }
-      return kvalues;
-    },
-
-    predict: function(classifierData, input) {
-      var nSupport = classifierData.n_support;
-      var supportVectors = classifierData.support_vectors;
-      var dualCoef = classifierData.dual_coef;
-      var intercept = classifierData.intercept;
-      var classes = classifierData.classes;
-      var kernelParams = classifierData.kernel_params;
-
-      var startIndices = [];
-      startIndices[0] = 0;
-      for (var i = 1; i < nSupport.length; i++) {
-        startIndices[i] = startIndices[i - 1] + nSupport[i - 1];
-      }
 
-      if (supportVectors[0].length !== input.length) {
-        // Support vector and input dimensions do not match.
-        $log.error(
-          'Dimension of support vectors and given input is different.');
-      }
+        var Qp = [];
+        for (var i = 0; i < nClasses; i++) {
+          Qp.push(0);
+        }
 
-      // Find kernel values for supportVectors and given input. Assumes that
-      // input has same dimension and data type as any of the supportVectors.
-      var kvalues = this.kernel(kernelParams, supportVectors, input);
+        var P = [];
+        for (var i = 0; i < nClasses; i++) {
+          P.push(0);
+        }
 
-      var votes = [];
-      for (var i = 0; i < classes.length; i++) {
-        votes.push(0);
-      }
+        var maxIter = Math.max(100, nClasses);
+        var eps = 0.005 / nClasses;
 
-      var p = 0;
-      for (var i = 0; i < classes.length; i++) {
-        for (var j = i + 1; j < classes.length; j++) {
-          var si = startIndices[i];
-          var sj = startIndices[j];
-          var ci = nSupport[i];
-          var cj = nSupport[j];
+        for (var t = 0; t < nClasses; t++) {
```

Comment:
I'm taking most of the changes here on trust; is there anything in particular you'd like me to look at here?

Also, if we're following an algorithm that's known, then could we add a link to the reference in case any reconstruction is needed in the future?

Reply:
I think I have added a link to its implementation in libsvm and also a link to the original research paper which proposed this algorithm.

Reply:
Ah yep -- see later comment below, sorry: https://github.com/oppia/oppia/pull/6012/files#r244641191